File: src/thread_load/disruptor.clj
(ns thread-load.disruptor
  (:import
    [thread_load Event]
    [java.util.concurrent Executors ExecutorService TimeUnit]
    [java.util.concurrent.atomic AtomicReference]
    [com.lmax.disruptor EventFactory EventHandler RingBuffer YieldingWaitStrategy BlockingWaitStrategy BusySpinWaitStrategy WaitStrategy]
    [com.lmax.disruptor.dsl Disruptor ProducerType ProducerType]))


(defn- ^EventFactory create-event-factory []
  (reify EventFactory
    (newInstance [this]
      (Event.))))


(defn- ^EventHandler event-handler [f]
  (reify EventHandler
    (onEvent [this evt seq endofbatch]
      (f (.getVal ^Event evt)))))


(defn publish! [{:keys [^RingBuffer ring-buffer]} data]
  (let [seq (.next ring-buffer)
        evt (.get ring-buffer seq)]
    (try
      (.setVal ^Event evt data)
      (finally
        (.publish ring-buffer seq)))))

(defn- get-producer-type [type]
  (cond
    (= type :multi) (ProducerType/MULTI)
    (= type :single) (ProducerType/SINGLE)
    :else
    (throw (RuntimeException. (str "Producer type must be :single or :multi but got " type)))))

(defn- ^WaitStrategy get-wait-strategy [type]
  (cond
    (= type :yield) (YieldingWaitStrategy.)
    (= type :block) (BlockingWaitStrategy.)
    (= type :spin)  (BusySpinWaitStrategy.)
    :else
    (throw (RuntimeException. (str "Wait strategy can only be :yield :block :spin but got " type)))))

(defn add-event-handler [{:keys [disruptor]} f]
  (.handleEventsWith ^Disruptor disruptor (into-array [(event-handler f)])))


(defn create-disruptor [event-f & {:keys [buffer-size producer-type wait-strategy executor-service]
                                   :or {buffer-size 1024 producer-type :multi wait-strategy :block executor-service nil}}]
  (let [^ExecutorService service (if executor-service executor-service (Executors/newCachedThreadPool))
        ^EventFactory event-factory (create-event-factory)
        ^Disruptor disruptor (doto
                                 (Disruptor. event-factory buffer-size
                                             service
                                             (get-producer-type producer-type)
                                             (get-wait-strategy wait-strategy))
                                 (.handleEventsWith (into-array [(event-handler event-f)]))
                                 .start)]
    {:exec-service service :ring-buffer (.getRingBuffer disruptor) :disruptor disruptor :exec-service-type (if executor-service :shared :unique) }))


(defn shutdown-pool [{:keys [disruptor exec-service-type exec-service]}]
  (.shutdown ^Disruptor disruptor)
  (when (= exec-service-type :unique)
     (.shutdown ^ExecutorService exec-service)
     (.awaitTermination ^ExecutorService exec-service 2000 TimeUnit/MILLISECONDS)
     (.shutdownNow ^ExecutorService exec-service)))

File: java/thread_load/disruptor/Disruptor.java
package thread_load.disruptor;

import clojure.lang.*;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Java support for thread-load.disruptor
 */
public class Disruptor {

    static {
        RT.var("clojure.core", "require").invoke(Symbol.create("thread-load.disruptor"));
    }

    private Var publishVar = RT.var("thread-load.disruptor", "publish!");

    private final Object disruptor;

    private Disruptor(Object disruptor){
        this.disruptor = disruptor;
    }

    public void publish(Object v){
        publishVar.invoke(disruptor, v);
    }

    public void shutdown(){
        RT.var("thread-load.disruptor", "shutdown-pool").invoke(disruptor);
    }

    public static Disruptor create(IFn f){
        return create(new DisruptorConf(), f);
    }

    public static Disruptor create(DisruptorConf conf, IFn f){
        IPersistentVector v =
                PersistentVector.EMPTY.cons(f);

        ISeq s = conf.getConf().seq();
        if(s != null){
            SeqIterator it = new SeqIterator(s);
            while(it.hasNext()) {
                MapEntry entry = (MapEntry) it.next();
                v = v.cons(entry.getKey()).cons(entry.getValue());
            }
        }

       return new Disruptor(RT.var("thread-load.disruptor", "create-disruptor").applyTo(v.seq()));
    }

    public static class DisruptorConf{
        public enum PRODUCER_TYPE {MULTI, SINGLE};
        public enum WAIT_STRATEGY {YIELD, BLOCK, SPIN};

        private static final Keyword KW_BUFFER_SIZE = Keyword.intern("buffer-size");
        private static final Keyword KW_PRODUCER_TYPE = Keyword.intern("producer-type");
        private static final Keyword KW_WAIT_STRATEGY = Keyword.intern("wait-strategy");
        private static final Keyword KW_EXEC_SERVICE = Keyword.intern("executor-service");

        IPersistentMap conf = PersistentArrayMap.EMPTY;

        public void setBufferSize(int bufferSize){
            conf = conf.assoc(KW_BUFFER_SIZE, bufferSize);
        }
        public void setProducerType(PRODUCER_TYPE type){
            if(type == PRODUCER_TYPE.MULTI)
                conf = conf.assoc(KW_PRODUCER_TYPE, Keyword.intern("multi"));
            else
                conf = conf.assoc(KW_PRODUCER_TYPE, Keyword.intern("single"));
        }
        public void setWaitStrategy(WAIT_STRATEGY s){
            if(s == WAIT_STRATEGY.BLOCK)
                conf = conf.assoc(KW_WAIT_STRATEGY, Keyword.intern("block"));
            else if(s == WAIT_STRATEGY.SPIN)
                conf = conf.assoc(KW_WAIT_STRATEGY, Keyword.intern("spin"));
            else
                conf = conf.assoc(KW_WAIT_STRATEGY, Keyword.intern("yield"));

        }

        public void setExecutorService(ExecutorService service){
            conf = conf.assoc(KW_EXEC_SERVICE, service);
        }

        public IPersistentMap getConf(){
            return conf;
        }
    }

    public static void test(){

        DisruptorConf conf = new DisruptorConf();
        conf.setExecutorService(Executors.newCachedThreadPool());
        conf.setProducerType(DisruptorConf.PRODUCER_TYPE.SINGLE);
        conf.setWaitStrategy(DisruptorConf.WAIT_STRATEGY.SPIN);

        Disruptor disruptor = Disruptor.create(conf,
           new AFn() {
            @Override
            public Object invoke(Object v) {
                System.out.println("Val1: " + v);
                return null;
            }
        });

        for(int i = 0; i < 10; i++)
            disruptor.publish(i);

        disruptor.shutdown();
    }
}
